source/opt/debug_info_manager.h

```diff
@@ -0,0 +1,132 @@
+  DebugInfoManager& operator=(const DebugInfoManager&) = delete;
+  DebugInfoManager& operator=(DebugInfoManager&&) = delete;
+
+  friend bool operator==(const DebugInfoManager&, const DebugInfoManager&);
+  friend bool operator!=(const DebugInfoManager& lhs,
+                         const DebugInfoManager& rhs) {
+    return !(lhs == rhs);
+  }
+
+  // Clones DebugDeclare or DebugValue for a local variable whose result id is
+  // |orig_var_id|. Set Variable operand of the new DebugDeclare as
+  // |new_var_id|. Return the new DebugDeclare or DebugValue. If this is the
+  // first DebugDeclare or DebugValue, keep it in |local_var_id_to_dbgdecl_|.
+  Instruction* CloneDebugDeclare(uint32_t orig_var_id, uint32_t new_var_id);
+
+  // Returns id of new DebugInlinedAt. Its Line operand is the line number
+  // of |line| if |line| is not nullptr. Otherwise, its Line operand
+  // is the line number of lexical scope of |scope|. Its Scope and Inlined
+  // operands are Scope and Inlined of |scope|. Note that this function puts
+  // the new DebugInlinedAt into the tail of debug instructions.
+  uint32_t CreateDebugInlinedAt(const Instruction* line,
+                                const DebugScope& scope);
+
+  // Creates a new DebugInfoNone instruction and returns it. In addition,
+  // insert the new DebugInfoNone instruction before the head of debug
+  // instructions.
+  Instruction* CreateDebugInfoNone();
+
+  // Returns DebugInlinedAt whose id is |dbg_inlined_at_id|. If it does not
+  // exist or it is not a DebugInlinedAt instruction, return nullptr.
+  Instruction* GetDebugInlinedAt(uint32_t dbg_inlined_at_id);
+
+  // Returns DebugFunction whose Function operand is |fn_id|. If it does not
+  // exist, return nullptr.
+  Instruction* GetDebugFunction(uint32_t fn_id) {
+    auto dbg_fn_it = fn_id_to_dbg_fn_.find(fn_id);
+    return dbg_fn_it == fn_id_to_dbg_fn_.end() ? nullptr : dbg_fn_it->second;
+  }
+
+  // Clones DebugInlinedAt whose id is |dbg_inlined_at_id|. If
+  // |dbg_inlined_at_id| is not an id of DebugInlinedAt, returns nullptr.
+  // Note that this function does not insert the new DebugInlinedAt into
+  // debug instruction list of the module.
+  Instruction* CloneDebugInlinedAt(uint32_t dbg_inlined_at_id);
+
+ private:
+  IRContext* context() { return context_; }
+
+  // Analyzes OpenCL.DebugInfo.100 instructions in the given |module| and
+  // populates data structures in this class.
+  void AnalyzeDebugInsts(Module& module);
+
+  // Returns the DebugDeclare instruction that corresponds to the variable
+  // with id |var_id|. Returns |nullptr| if one does not exists.
+  Instruction* GetDbgDeclareForVar(uint32_t var_id);
+
+  // Registers the DebugDeclare instruction |inst| into
+  // |local_var_id_to_dbgdecl_| using the variable operand of |inst| as a key.
+  // If |local_var_id_to_dbgdecl_| already has a key that corresponds to the
+  // variable operand of |inst|, it just returns without doing anything.
```

Comment:
Shouldn't it be an error if the key already exists with a different instruction?  Why would this not be a sign that something is wrong?

I believe that you do this because you want to clone debug declare instruction and register them right away.  I think this shows that it might be the wrong direction.  It should be up to inlining to first clone the function.  Update the variable to the new variable, and then you register to the new debug declare instruction.

This shows why we need a public function to register new debug instructions.  We do not have all of the necessary information when doing the cloning.